File: javascript/json-transform/src/functions/TransformerFunctionDecompress.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import TextEncoding from "./common/TextEncoding";

class TransformerFunctionDecompress extends TransformerFunction {
  constructor() {
    super({
      argsSet: [
        { name: "charset", type: ArgType.String, defaultValue: "ISO-8859-1" },
        { name: "format", type: ArgType.String, defaultValue: "UTF-8" },
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    const charset = await context.getEnum("charset");
    if (!charset) return null;
    const format = await context.getEnum("format");
    if (!format) return null;

    const bytes = TextEncoding.encode(str, charset);
    const stream = new Blob([bytes], { type: "text/plain" }).stream();
    const decompressedStream = stream.pipeThrough(new DecompressionStream("gzip"));
    const response = new Response(decompressedStream);
    const outBytes = await response.blob().then(b => b.arrayBuffer());
    return TextEncoding.decode(new Uint8Array(outBytes), format);
  }
}

export default TransformerFunctionDecompress;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionDecompress.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;

import java.io.*;
import java.util.Map;
import java.util.zip.GZIPInputStream;

public class TransformerFunctionDecompress extends TransformerFunction {
    public TransformerFunctionDecompress() {
        super(FunctionDescription.of(
                Map.of(
                        "charset", ArgumentType.of(ArgType.String).position(0).defaultValue("ISO-8859-1"),
                        "format", ArgumentType.of(ArgType.String).position(1).defaultValue("UTF-8")
                )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var charset = context.getEnum("charset");
        if (charset == null) {
            return null;
        }
        var format = context.getEnum("format");
        if (format == null) {
            return null;
        }

        try (var bais = new ByteArrayInputStream(str.getBytes(charset));
             var gzip = new GZIPInputStream(bais);
             var isr = new InputStreamReader(gzip, format);
             var sw = new StringWriter()) {

            char[] buffer = new char[8192];
            int len;
            while ((len = isr.read(buffer)) > 0) {
                sw.write(buffer, 0, len);
            }
            return sw.toString();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}
